modules/lang-painless/src/main/java/org/elasticsearch/painless/Definition.java

```diff
@@ -33,13 +33,16 @@
 import java.util.Map;
 import java.util.Objects;
 import java.util.Stack;
+import java.util.regex.Pattern;
 
 /**
  * The entire API for Painless.  Also used as a whitelist for checking for legal
  * methods and fields during at both compile-time and runtime.
  */
 public final class Definition {
 
+    private static final Pattern TYPE_NAME_PATTERN = Pattern.compile("^[_a-zA-Z][._a-zA-Z0-9]*$");
```

Comment:
A type name can start with an underscore?

Reply:
Yes.  This matches the ANTLR grammar.  This can probably be changed, but I don't want to add more complications to grammar ID's right now.